### PR TITLE
Docs for customizable Admin components

### PR DESCRIPTION
**Pull request by jamesricky**
---

## PR Checklist

-   [x] Link to the respective task if one exists: COM-647
-   [x] Link to the deployed docs that are affected:
    - [Admin Components: Customizaton and Styling](https://deploy-preview-2030--comet-dxp-docs.netlify.app/docs/admin-components/customization-and-styling)
    - [COMET Core Development: Creating customizable admin components](https://deploy-preview-2030--comet-dxp-docs.netlify.app/docs/comet-core-development/creating-customizable-admin-components)

### CODE REVIEW

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
+---
+title: Creating customizable admin components
```
Comment (johnnyomair) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:2]:
```suggestion
title: Creating customizable Admin components
```

IMO Admin should be uppercase

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+The outermost slot of a component is generally called `root`.
+
+```tsx
+export type MyComponentClassKey = "root" | "title";
+```
+
+## Slots
+
+Each element or sub-component of a Comet Admin component is defined as a slot.
```
Comment (johnnyomair) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:37]:
```suggestion
Each element or subcomponent of a Comet Admin component is defined as a slot.
```

I believe sub is usually used without a hyphen

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+By default, a slots class name consists of the component and slot names, prefixed with `CometAdmin`.
+For example, the `root` slot of `MyComponent` would have the class name `CometAdminMyComponent-root`, while the `title` slot would have the class name `CometAdminMyComponent-title`.
+
+Additional class names can be added to a slot by returning an array of class keys in `classesResolver()`.
+This allows easy customization using its class name in a selector outside the component.
+
+In the following example, a component has a conditional `highlighted` state.
+If `true`, the class name `CometAdminMyComponent-highlighted` is added to the `root` slot by returning the `highlighted` class key in `classesResolver`.
+If `false`, `classesResolver` returns an empty array, so no additional class name is added.
```
Comment (johnnyomair) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:154]: This isn't true. The array isn't empty, but `[false]` instead. Could this lead to problems?

image

File: docs/docs/admin-components/customization-and-styling.mdx [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+import { Story } from "../../src/components/Story";
+
+Comet Admin components are designed to be used similarly to MUI components.
+The APIs for styling, theming, overriding props, and replacing icons are similar to MUI.
+
+Following are examples of customizing the `ContentOverflow` component.
+The examples should all be transferable to other components.
+
+For instructions on how to build components that support these features, see [Creating customizable admin components](/docs/comet-core-development/creating-customizable-admin-components).
```
Comment (johnnyomair) [inline docs/docs/admin-components/customization-and-styling.mdx:14]:
```suggestion
For instructions on how to build components that support these features, see [Creating customizable Admin components](/docs/comet-core-development/creating-customizable-admin-components).
```

**Review by johnnyomair — COMMENTED** [review]
Thanks for this in-depth documentation! It's easy to understand 👍🏼

File: docs/docs/comet-core-development/creating-customizable-admin-components.md (hunk 3 above)
Comment (jamesricky) [reply docs/docs/comet-core-development/creating-customizable-admin-components.md:154]: Good point, will re-write this. 

> Could this lead to problems?

No, `falsy` values are filtered out:

https://github.com/vivid-planet/comet/blob/89092f9e2b983cc689d6d2238d46ec4930992aa9/packages/admin/admin/src/helpers/createComponentSlot.tsx#L38

**Review by johnnyomair — APPROVED** [review]

File: docs/docs/admin-components/customization-and-styling.mdx [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+## Commonly used term: Slot
+
+A component is made up of HTML elements and other components.
+Each of these elements or components is referred to as a slot.
+Each of the component's slots can be accessed through its name to override styles and props.
+The outermost slot of a component is generally called `root`.
+
+## Customizing individial instances of a component
+
+### Styling a components root slot
```
Comment (thomasdax98) [inline docs/docs/admin-components/customization-and-styling.mdx:25]:
```suggestion
### Styling a component's root slot
```

File: docs/docs/admin-components/customization-and-styling.mdx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+>
+    {/* ... */}
+</ContentOverflow>
+```
+
+<Story path="customization-and-styling/OverridingThePropsOfASlot.tsx" />
+
+### Styling a slot
+
+Overriding the styles of a components slot can be done by accessing its `sx` prop through the `slotProps` prop.
```
Comment (thomasdax98) [inline docs/docs/admin-components/customization-and-styling.mdx:60]:
```suggestion
Overriding the styles of a component's slot can be done by accessing its `sx` prop through the `slotProps` prop.
```

File: docs/docs/admin-components/customization-and-styling.mdx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        },
+    },
+});
+```
+
+<Story path="customization-and-styling/OverridingStylesOfAComponentGlobally.tsx" />
+
+### Setting default props of a component
+
+Overriding the default props of a component can be done in the theme using the components `defaultProps`.
```
Comment (thomasdax98) [inline docs/docs/admin-components/customization-and-styling.mdx:110]:
```suggestion
Overriding the default props of a component can be done in the theme using the component's `defaultProps`.
```

File: docs/docs/admin-components/customization-and-styling.mdx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+</ContentOverflow>
+```
+
+<Story path="customization-and-styling/StylingASlot.tsx" />
+
+## Customizing components globally
+
+### Overriding styles of a component
+
+Overriding the styles of a component can be done in the theme using the components `styleOverrides`.
```
Comment (thomasdax98) [inline docs/docs/admin-components/customization-and-styling.mdx:85]:
```suggestion
Overriding the styles of a component can be done in the theme using the component's `styleOverrides`.
```

File: docs/docs/admin-components/customization-and-styling.mdx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    },
+});
+```
+
+<Story path="customization-and-styling/OverridingStylesOfAComponentGlobally.tsx" />
+
+### Setting default props of a component
+
+Overriding the default props of a component can be done in the theme using the components `defaultProps`.
+In this case, we replace the "openDialog" icon using the `iconMapping` prop.
```
Comment (thomasdax98) [inline docs/docs/admin-components/customization-and-styling.mdx:111]:
```suggestion
In this case, we replace the `openDialog` icon using the `iconMapping` prop.
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+title: Creating customizable Admin components
+---
+
+MUI components support multiple ways of customization.
+You can override the styles of individual component instances or style them globally using the theme.
+You can also use the theme to override those components' default props.
+
+All Comet Admin components should also support these features.
+
+For examples of how those features should be able to be used, see the [Customization and Styling](/docs/admin-components/customization-and-styling).
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:11]:
```suggestion
Those features should be useable as described in [Customization and Styling](/docs/admin-components/customization-and-styling).
```

Maybe like this?

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+-   The props type, named as the component, followed by `Props`
+-   The class key type, named as the component, followed by `ClassKey`
+
+```tsx
+export { MyComponent, MyComponentProps, MyComponentClassKey } from "./src/MyComponent";
+```
+
+## Class keys
+
+A components `ClassKey` type defines all the individual elements (or slots) of a component or a modifier/alternative style of a component.
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:27]:
```suggestion
A component's `ClassKey` type defines all the individual elements (or slots) of a component and all modifiers/alternative styles of a component.
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+```tsx
+export type MyComponentClassKey = "root" | "title";
+```
+
+## Slots
+
+Each element or subcomponent of a Comet Admin component is defined as a slot.
+
+A slot is created and styled by using `createComponentSlot()` and passing in the HTML element or component you want to be the base of your slot.
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:39]:
```suggestion
A slot is created and styled by using `createComponentSlot()` and passing in the HTML element or component you want to base your slot on.
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    }> {
+    variant?: "primary" | "secondary";
+    children?: React.ReactNode;
+}
+```
+
+## Using props
+
+Props should not be used directly but through the `useThemeProps` hook instead.
+This ensures that the components `defaultProps` from the theme are merged with the passed-in props.
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:80]:
```suggestion
This ensures that the component's `defaultProps` from the theme are merged with the passed-in props.
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+Additionally, the `sx` and `className` props should be passed into the root slot, for example with `...restProps`.
+
+```tsx
+export function MyComponent(inProps: MyComponentProps) {
+    const { slotProps, ...restProps } = useThemeProps({
+        props: inProps,
+        name: "CometAdminMyComponent",
+    });
+    return (
+        <Root {...slotProps?.root} {...restProps}>
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:105]: Should the slotProps always be the first prop?

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+            <Title {...slotProps?.title} />
+        </Root>
+    );
+}
+```
+
+## Owner state
+
+The `ownerState` is an object that includes all values that should impact the component's styling.
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:114]:
```suggestion
The `ownerState` is an object that includes all values that can impact the component's styling.
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+    // ...
+    const ownerState: OwnerState = {
+        highlighted: getHighlightedState(),
+    };
+
+    return <Root ownerState={ownerState} {...slotProps?.root} {...restProps} />;
+}
+```
+
+## A slots `classesResolver`
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:144]:
```suggestion
## A slot's `classesResolver`
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+
+By default, a slots class name consists of the component and slot names, prefixed with `CometAdmin`.
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:146]:
```suggestion
By default, a slot's class name consists of the component and slot names, prefixed with `CometAdmin`.
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+For example, the `root` slot of `MyComponent` would have the class name `CometAdminMyComponent-root`, while the `title` slot would have the class name `CometAdminMyComponent-title`.
+
+Additional class names can be added to a slot by returning an array of class keys in `classesResolver()`.
+This allows easy customization using its class name in a selector outside the component.
```
Comment (thomasdax98) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:150]:
```suggestion
This allows easy customization using the resulting class name in a selector outside the component.
```

File: docs/docs/comet-core-development/creating-customizable-admin-components.md (hunk 14 above)
Comment (jamesricky) [reply docs/docs/comet-core-development/creating-customizable-admin-components.md:105]: Generally `slotProps` and `restProps` should be last, so props defined inside the component can be overridden. 
Exceptions would be props that should never be overridden. 

I will add this info to the docs as well 👍🏼

File: docs/docs/comet-core-development/creating-customizable-admin-components.md [hunk 19]
```diff
@@ -8,7 +8,7 @@ You can also use the theme to override those components' default props.
 
 All Comet Admin components should also support these features.
 
-For examples of how those features should be able to be used, see the [Customization and Styling](/docs/admin-components/customization-and-styling).
+Those features should be useable as described in [Customization and Styling](/docs/admin-components/customization-and-styling).
```
Comment (johnnyomair) [inline docs/docs/comet-core-development/creating-customizable-admin-components.md:11]:
```suggestion
Those features should be usable as described in [Customization and Styling](/docs/admin-components/customization-and-styling).
```

Shouldn't this be usable?

**Review by johnnyomair — APPROVED** [review]

**Review by thomasdax98 — APPROVED** [review]

**Pull request merged**